### DISCUSSION TITLE
EXCLUDE_FROM_ALL

### DISCUSSION THREAD

**Discussion by flagarde** [cpm-cmake/CPM.cmake#242] [Q&A]
Hello,

Reading the doc it appears that `All packages added using the shorthand syntax will be added using the EXCLUDE_FROM_ALL flag.`. Could you explain the reason of such behaviour?

**Comment by TheLartians** [comment]
Hey @flagarde, thanks for the question! Please see the [relevant issue](https://github.com/cpm-cmake/CPM.cmake/issues/152) for a detailed discussion on that topic.

**TL;DR:** CMake libraries are often not made to be imported directly (without installation) and often define a bunch of useless targets that get included in the `all` target. To make the shorthand syntax compatible with such projects, we decided to make `EXCLUDE_FROM_ALL` the default behaviour.

One issue that we didn't consider is that CMake still adds tests, which can potentially break the `tests` taget. Imho this behaviour is a CMake bug and an [issue has already been raised](https://gitlab.kitware.com/cmake/cmake/-/issues/20212) on their tracker.

**Reply by flagarde** [reply]
Thx to pointing this out. I understand the benefics of this option but my question was more like : Why the shorthand suppose you want EXCLUDE_FROM_ALL by default and the standard doesn't. I understood you didn't wanted to break old behavior but it makes the shorthand different from the other one and from CMake which set it to ON by default. It's fine as it's written on the doc but I was you trying to guess why you choose this option. Then if I understand correctly is not possible to pass options to the shorthand so it's impossibe to set EXCLUDE_FROM_ALL to ON ?

**Reply by iboB** [reply]
The shorthand syntax was added after the original syntax when it was decided that automatically setting exclude EXCLUDE_FROM_ALL was a better default. And yes, since int's impossible to pass options with the shorthand, it now sets it to ON automatically.

**Reply by flagarde** [reply]
@iboB Thx !
